File: src/lcpush/plausibility.py
"""Clipboard plausibility scoring (spec §6.2).

This decides *menu ordering only*. It never blocks a choice, and the clipboard
stays selectable however badly it scores.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from .detect import detect
from .solution import LARGE_CLIPBOARD_BYTES, byte_size

PLAUSIBLE_THRESHOLD = 3

ENTRY_SHAPES = re.compile(
    r"class\s+Solution|impl\s+Solution|object\s+Solution|public\s+class\s+Solution"
    r"|var\s+\w+\s*=\s*function\s*\(|func\s+\w+\s*\(|def\s+\w+\s*\(|fun\s+\w+\s*\(",
    re.MULTILINE,
)

BARE_URL = re.compile(r"^\s*(https?://|www\.)\S+\s*$", re.IGNORECASE)
BARE_EMAIL = re.compile(r"^\s*[\w.+-]+@[\w-]+\.[\w.]+\s*$")
BARE_PATH = re.compile(r"^\s*(~|\.{0,2}/|[A-Za-z]:\\)[\w./\\ -]*\s*$")
WORD = re.compile(r"[A-Za-z']+")
SYMBOL = re.compile(r"[{}()\[\];:=<>+*/&|%!#$@\\]")


@dataclass(frozen=True)
class Plausibility:
    plausible: bool
    score: int
    reasons: tuple[str, ...] = field(default=())


def _prose_heavy(text: str) -> bool:
    """High word-to-symbol ratio: the signature of chat messages and articles."""
    words = len(WORD.findall(text))
    symbols = len(SYMBOL.findall(text))
    if words < 12:
        return False
    return symbols == 0 or words / max(symbols, 1) > 12


def assess(text: str | None) -> Plausibility:
    """Score clipboard content as "looks like a LeetCode solution" or not."""
    if not text or not text.strip():
        return Plausibility(False, 0, ("clipboard is empty",))

    points = 0
    reasons: list[str] = []
    lines = text.strip().splitlines()

    if detect(text).confident:
        points += 3
        reasons.append("language detected")
    if ENTRY_SHAPES.search(text):
        points += 3
        reasons.append("has a solution entry point")
    if len(lines) >= 2:
        points += 1
        reasons.append("multi-line")
    if SYMBOL.search(text) and any(
        line[:1] in (" ", "\t") for line in lines if line.strip()
    ):
        points += 1
        reasons.append("indented code-like punctuation")

    single_line = text.strip()
    if len(lines) == 1 and (
        BARE_URL.match(single_line)
        or BARE_EMAIL.match(single_line)
        or BARE_PATH.match(single_line)
    ):
        points -= 5
        reasons.append("looks like a bare URL, email, or path")
    if len(lines) == 1 and len(single_line) < 40:
        points -= 3
        reasons.append("single short line")
    if _prose_heavy(text):
        points -= 3
        reasons.append("reads as prose")
    if byte_size(text) > LARGE_CLIPBOARD_BYTES:
        points -= 3
        reasons.append("over 200KB")

    return Plausibility(points >= PLAUSIBLE_THRESHOLD, points, tuple(reasons))

File: src/lcpush/domain/plausibility.cpp
#include "lcpush/domain/plausibility.hpp"

#include <regex>

#include "lcpush/domain/detect.hpp"
#include "lcpush/domain/solution.hpp"
#include "lcpush/util/strings.hpp"

namespace lcpush::plausibility {

namespace {

const std::regex& entry_shapes() {
    static const std::regex value(
        R"(class\s+Solution|impl\s+Solution|object\s+Solution|public\s+class\s+Solution)"
        R"(|var\s+\w+\s*=\s*function\s*\(|func\s+\w+\s*\(|def\s+\w+\s*\(|fun\s+\w+\s*\()",
        std::regex::ECMAScript | std::regex::multiline);
    return value;
}

const std::regex& bare_url() {
    static const std::regex value(R"(^\s*(https?://|www\.)\S+\s*$)",
                                  std::regex::ECMAScript | std::regex::icase);
    return value;
}

const std::regex& bare_email() {
    static const std::regex value(R"(^\s*[\w.+-]+@[\w-]+\.[\w.]+\s*$)");
    return value;
}

const std::regex& bare_path() {
    static const std::regex value(R"(^\s*(~|\.{0,2}/|[A-Za-z]:\\)[\w./\\ -]*\s*$)");
    return value;
}

const std::regex& word() {
    static const std::regex value(R"([A-Za-z']+)");
    return value;
}

const std::regex& symbol() {
    static const std::regex value(R"([{}()\[\];:=<>+*/&|%!#$@\\])");
    return value;
}

size_t count_matches(const std::string& text, const std::regex& pattern) {
    size_t count = 0;
    for (auto it = std::sregex_iterator(text.begin(), text.end(), pattern);
         it != std::sregex_iterator(); ++it) {
        ++count;
    }
    return count;
}

// High word-to-symbol ratio: the signature of chat messages and articles.
bool prose_heavy(const std::string& text) {
    size_t words = count_matches(text, word());
    size_t symbols = count_matches(text, symbol());
    if (words < 12) return false;
    if (symbols == 0) return true;
    return static_cast<double>(words) / static_cast<double>(symbols) > 12.0;
}

// Non-empty logical lines of the trimmed text, like Python str.splitlines
// on the inputs this sees (LF-normalized clipboard text).
std::vector<std::string> split_lines(const std::string& text) {
    return util::split(text, '\n');
}

}  // namespace

Plausibility assess(const std::optional<std::string>& text) {
    if (!text || util::trim(*text).empty()) {
        return Plausibility{false, 0, {"clipboard is empty"}};
    }

    int points = 0;
    std::vector<std::string> reasons;
    std::string stripped = util::trim(*text);
    std::vector<std::string> lines = split_lines(stripped);

    if (detect::detect(*text).confident()) {
        points += 3;
        reasons.push_back("language detected");
    }
    if (std::regex_search(*text, entry_shapes())) {
        points += 3;
        reasons.push_back("has a solution entry point");
    }
    if (lines.size() >= 2) {
        points += 1;
        reasons.push_back("multi-line");
    }
    bool has_indent = false;
    for (const std::string& line : lines) {
        if (!util::trim(line).empty() && (line.rfind(' ', 0) == 0 || line.rfind('\t', 0) == 0)) {
            has_indent = true;
            break;
        }
    }
    if (std::regex_search(*text, symbol()) && has_indent) {
        points += 1;
        reasons.push_back("indented code-like punctuation");
    }

    if (lines.size() == 1 &&
        (std::regex_match(stripped, bare_url()) || std::regex_match(stripped, bare_email()) ||
         std::regex_match(stripped, bare_path()))) {
        points -= 5;
        reasons.push_back("looks like a bare URL, email, or path");
    }
    if (lines.size() == 1 && util::codepoint_count(stripped) < 40) {
        points -= 3;
        reasons.push_back("single short line");
    }
    if (prose_heavy(*text)) {
        points -= 3;
        reasons.push_back("reads as prose");
    }
    if (solution::byte_size(*text) > solution::kLargeClipboardBytes) {
        points -= 3;
        reasons.push_back("over 200KB");
    }

    return Plausibility{points >= kPlausibleThreshold, points, std::move(reasons)};
}

}  // namespace lcpush::plausibility
